### ISSUE TITLE
MacOS M3 gpu Trainer error with torchmetrics top k accuracy

### ISSUE THREAD

**Issue by qxakshat** [Lightning-AI/pytorch-lightning#19566]
### Bug description

I am using python 3.11.5 torch == 2.2.1

getting 
`/AppleInternal/Library/BuildRoots/4e1473ee-9f66-11ee-8daf-cedaeb4cabe2/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArrayScatter.mm:308: failed assertion `Rank of destination array (1) must be greater than or equal to inner-most dimension of indices array (2)'`

with torch metrics top k accuracy metric on MAC M3 GPU. No error with CPU.
Sample code in attached,

### What version are you seeing the problem on?

v2.2, master

### How to reproduce the bug

```python
#!/usr/bin/env python
# coding: utf-8

import multiprocessing
multiprocessing.set_start_method('fork')

import torch
import torch.nn as nn
import torch.functional as F

import torchmetrics
import torchvision

import lightning as L
from lightning import Trainer, LightningModule

transform = torchvision.models.MobileNet_V3_Small_Weights.IMAGENET1K_V1.transforms()
train_dataset = torchvision.datasets.CIFAR100(root='./data/', transform=transform, train = True)
train_dataloader = torch.utils.data.DataLoader(train_dataset, batch_size=256, shuffle=True, num_workers=8, persistent_workers=True)


# define the LightningModule
class LightningModule(L.LightningModule):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.loss = nn.CrossEntropyLoss()
        
        self.metric_top1 = torchmetrics.classification.Accuracy(task="multiclass", num_classes=100)
        self.metric_top2 = torchmetrics.classification.Accuracy(task="multiclass", num_classes=100, top_k=2)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.model(x)
        loss = self.loss(y_hat, y)
        metrics = self.metric_top1(y_hat, y)
        mtop2 = self.metric_top2(y_hat, y)       < --- error here 
        self.log("train_loss", loss, on_epoch=True, prog_bar=True, logger=True)
        return loss
        
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

m = torch.hub.load('pytorch/vision', 'mobilenet_v3_small', weights=None, num_classes=100)
model = LightningModule(m)

trainer = Trainer(max_epochs=1, accelerator="gpu") < ---- working fine with "cpu"
trainer.fit(model, train_dataloader)
```

### Error messages and logs

```
/AppleInternal/Library/BuildRoots/4e1473ee-9f66-11ee-8daf-cedaeb4cabe2/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArrayScatter.mm:308: failed assertion `Rank of destination array (1) must be greater than or equal to inner-most dimension of indices array (2)'
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

>>> torchmetrics.__version__
'1.3.1'
>>> torchvision.__version__
'0.17.1'

**Comment by awaelchli** [comment]
I was able to isolate the issue to what looks like a PyTorch bug:
https://github.com/pytorch/pytorch/issues/121557

Feel free to follow the linked issue for updates.
